text/0477-policy-validation.md

```diff
@@ -0,0 +1,544 @@
+# CDK Policy validation
+
+* **Original Author(s):**: @otaviomacedo, @corymhall
+* **Tracking Issue**: <https://github.com/aws/aws-cdk-rfcs/issues/477>
+* **API Bar Raiser**: @iliapolo
+
+Allows users to validate that their CDK application complies with infrastructure policies at synthesis time. Integrates
+with third party policy-as-code tools, such as CloudFormation Guard, OPA and Checkov.
+
+## Working Backwards
+
+### Policy Validation
+
+It is possible to use policy as code tools such
+as [CloudFormation Guard](https://docs.aws.amazon.com/cfn-guard/latest/ug/what-is-guard.html)
+or [OPA](https://www.openpolicyagent.org/) to evaluate the compliance of CDK applications.
+
+Policy as code tools are integrated with CDK through a plugin mechanism so in order to add policy enforcement for a
+specific tool you need to first specify the plugin to use. You can do this by importing the plugin and applying it to a
+stage. For example, to add a hypothetical CloudFormation Guard validation to a stage you would do the following:
+
+```ts
+// globally for the entire app (an app is a stage)
+import { CfnGuardValidator } from '@aws-cdk/cfn-guard-validator';
+
+const app = new App({
+  validationPlugins: [
+    new CfnGuardValidator({
+      rules: [
+        // rules can be specified inline
+        Rules.fromAsset('../my-local-rules'),
+        
+        // or from a remote location
+        Rules.fromDownload('https://somelocation.com/company-rules', {
+          auth: {},
+        }),
+        
+        // or from an S3 location
+        Rules.fromS3(s3Location),
+        
+        // or by calling an external endpoint
+        Rules.fromRequest('POST', 'https://someendpoint.com/validate'),
+      ],
+      options: {
+        someOption: "value",
+      },
+    }),
+  ],
+});
+
+// only apply to a particular stage
+const prodStage = new Stage(app, 'ProdStage', {
+  validationPlugins: [...],
+});
+```
+
+The details of how to configure the plugin will be specific to each plugin. For example, the `CfnGuardValidator` shown
+above has a `rules` property which is an array of `Rules` objects. The `Rules` class is a helper class that can be used
+to specify where the rules are located.
+
+CDK Policy Validation should never be used as a deployment gate or as the enforcement layer. It should instead be
```

Comment:
Yeah we should add some more explanation here.  We're trying to say that you should not rely on this feature in order to prevent the deployment of misconfigured infrastructure. It's fine if it runs as part of a deployment pipeline and fails the pipeline if it is not compliant, but it should be viewed as more of "fail fast(er)" tool rather than a preventative measure. I guess both could be viewed as a type of "gate" though.

Reply:
I've added a note.